// @2008muyu/pi-ui-prompt-dialog/dist/index.js
// src/prompt-dialog.ts
import {
  CURSOR_MARKER,
  matchesKey
} from "@earendil-works/pi-tui";
function visLen(s) {
  let len = 0;
  const stripEsc = /\x1b\[[0-9;]*m/g;
  for (const ch of s.replace(stripEsc, "")) {
    const code = ch.codePointAt(0);
    if (code >= 4352 && code <= 4447 || // Hangul Jamo
    code >= 11904 && code <= 42191 || // CJK
    code >= 44032 && code <= 55203 || // Hangul Syllables
    code >= 63744 && code <= 64255 || // CJK Compatibility
    code >= 65072 && code <= 65103 || // CJK Compatibility Forms
    code >= 65281 && code <= 65376 || // Fullwidth Forms
    code >= 127744 && code <= 129535) {
      len += 2;
    } else {
      len += 1;
    }
  }
  return len;
}
var PromptDialogComponent = class {
  width = 70;
  /** Focusable interface — set by TUI when focus changes */
  focused = false;
  selected = 0;
  actions;
  inputConfig;
  inputText = "";
  inputCursor = 0;
  inputFocused = false;
  title;
  description;
  theme;
  done;
  /** Total navigation items: actions + (optional) input row */
  get itemCount() {
    return this.actions.length + (this.inputConfig ? 1 : 0);
  }
  constructor(options, theme, done) {
    this.theme = theme;
    this.done = done;
    this.title = options.title;
    this.description = options.description || "";
    this.actions = options.actions;
    this.inputConfig = options.input;
    const defIdx = this.actions.findIndex((a) => a.default);
    this.selected = defIdx >= 0 ? defIdx : 0;
  }
  handleInput(data) {
    if (matchesKey(data, "escape")) {
      if (this.inputFocused) {
        this.inputFocused = false;
      } else {
        this.done(void 0);
      }
      return;
    }
    if (matchesKey(data, "return")) {
      if (this.inputFocused) {
        this.done({ type: "input", value: this.inputText });
      } else if (this.inputConfig && this.selected === this.actions.length) {
        this.inputFocused = true;
      } else {
        const action = this.actions[this.selected];
        this.done({ type: "action", action: action.id });
      }
      return;
    }
    if (this.inputConfig && (data === "	" || matchesKey(data, "tab"))) {
      if (!this.inputFocused) {
        this.inputFocused = true;
        return;
      }
    }
    if (!this.inputFocused) {
      if (matchesKey(data, "up")) {
        this.selected = Math.max(0, this.selected - 1);
        return;
      }
      if (matchesKey(data, "down")) {
        this.selected = Math.min(this.itemCount - 1, this.selected + 1);
        return;
      }
      if (this.inputConfig && this.selected === this.actions.length && matchesKey(data, "right")) {
        this.inputFocused = true;
        return;
      }
    }
    if (this.inputFocused) {
      if (matchesKey(data, "left")) {
        this.inputCursor = Math.max(0, this.inputCursor - 1);
        return;
      }
      if (matchesKey(data, "right")) {
        this.inputCursor = Math.min(this.inputText.length, this.inputCursor + 1);
        return;
      }
      if (matchesKey(data, "backspace")) {
        if (this.inputCursor > 0) {
          this.inputText = this.inputText.slice(0, this.inputCursor - 1) + this.inputText.slice(this.inputCursor);
          this.inputCursor--;
        }
        return;
      }
      if (matchesKey(data, "up")) {
        this.inputFocused = false;
        this.selected = this.actions.length;
        return;
      }
      if (data.length === 1 && data.codePointAt(0) >= 32) {
        this.inputText = this.inputText.slice(0, this.inputCursor) + data + this.inputText.slice(this.inputCursor);
        this.inputCursor++;
        return;
      }
    }
  }
  render(_width) {
    const w = this.width;
    const th = this.theme;
    const innerW = w - 2;
    const lines = [];
    const pad = (s, len) => {
      const vis = visLen(s);
      return s + " ".repeat(Math.max(0, len - vis));
    };
    const row = (content) => th.fg("border", "\u2502") + pad(content, innerW) + th.fg("border", "\u2502");
    const inputRowIndex = this.actions.length;
    lines.push(th.fg("border", `\u256D${"\u2500".repeat(innerW)}\u256E`));
    lines.push(row(` ${th.fg("accent", th.bold ? th.bold(this.title) : this.title)}`));
    if (this.description) {
      for (const descLine of this.description.split("\n")) {
        if (descLine.trim()) {
          lines.push(row(` ${th.fg("muted", descLine)}`));
        }
      }
    }
    lines.push(row(` ${th.fg("dim", "\u2500".repeat(innerW - 2))}`));
    for (let i = 0; i < this.actions.length; i++) {
      const action = this.actions[i];
      const isSelected = i === this.selected && !this.inputFocused;
      const prefix = isSelected ? " \u25B6 " : "   ";
      const marker = action.default ? " * " : "";
      const label = isSelected ? th.fg("accent", `${marker}${action.label}`) : th.fg("text", `${marker}${action.label}`);
      lines.push(row(`${prefix}${label}`));
    }
    if (this.inputConfig) {
      const isSelected = this.selected === inputRowIndex && !this.inputFocused;
      const activeLine = isSelected || this.inputFocused;
      const prefix = isSelected ? " \u25B6 " : "   ";
      const labelText = activeLine ? th.fg("accent", `${this.inputConfig.label}:`) : th.fg("muted", `${this.inputConfig.label}:`);
      let inputDisplay;
      if (this.inputFocused) {
        const before = this.inputText.slice(0, this.inputCursor);
        const cursorChar = this.inputCursor < this.inputText.length ? this.inputText[this.inputCursor] : " ";
        const after = this.inputText.slice(this.inputCursor + 1);
        const marker = this.focused ? CURSOR_MARKER : "";
        inputDisplay = `${before}${marker}\x1B[7m${cursorChar}\x1B[27m${after}`;
      } else {
        inputDisplay = this.inputText || th.fg("dim", this.inputConfig.placeholder || "");
      }
      lines.push(row(`${prefix}${labelText} ${inputDisplay}`));
    }
    const help = this.inputFocused ? `\u2190 \u2192 move cursor \xB7 \u232B backspace \xB7 \u2191 exit \xB7 Esc cancel \xB7 Enter submit` : `\u2191\u2193 navigate \xB7 Tab input \xB7 Enter select \xB7 Esc cancel`;
    lines.push(row(""));
    lines.push(row(` ${th.fg("dim", help)}`));
    lines.push(th.fg("border", `\u2570${"\u2500".repeat(innerW)}\u256F`));
    return lines;
  }
  invalidate() {
  }
  dispose() {
  }
};

// src/types.ts
function validateOptions(options) {
  const { actions } = options;
  if (!actions || actions.length === 0) {
    throw new Error("At least one action is required.");
  }
  if (actions.length > 10) {
    throw new Error("Max 10 actions allowed.");
  }
  const ids = /* @__PURE__ */ new Set();
  for (const action of actions) {
    if (!action.id) throw new Error("Each action must have an id.");
    if (ids.has(action.id)) throw new Error(`Duplicate action id: "${action.id}".`);
    ids.add(action.id);
  }
  const defaults = actions.filter((a) => a.default).length;
  if (defaults > 1) throw new Error("Only one action can have default: true.");
}

// src/index.ts
async function showPromptDialog(ctx, options) {
  validateOptions(options);
  if (!ctx.hasUI) {
    throw new Error(
      "showPromptDialog() requires an interactive UI (ctx.hasUI). In headless / RPC mode, use ctx.ui.select() or ctx.ui.input() directly."
    );
  }
  const result = await ctx.ui.custom(
    (_tui, theme, _keybindings, done) => new PromptDialogComponent(options, theme, done)
  );
  return result;
}
export {
  PromptDialogComponent,
  showPromptDialog,
  validateOptions
};
